pandas/core/indexes/datetimes.py

```diff
@@ -407,12 +407,13 @@ def _maybe_utc_convert(self, other: Index) -> Tuple["DatetimeIndex", Index]:
         this = self
 
         if isinstance(other, DatetimeIndex):
-            if self.tz is not None:
-                if other.tz is None:
-                    raise TypeError("Cannot join tz-naive with tz-aware DatetimeIndex")
-            elif other.tz is not None:
+            if (
+                (self.tz is not None)
+                and (other.tz is None)
+                or (self.tz is None)
+                and (other.tz is not None)
+            ):
```

Comment:
`(self.tz is None) ^ (other.tz is None)`

Reply:
`(self.tz is None ^ other.tz is not None)`? AFAIKT the check is to see if one of them is `None` but the other one isn't - I've tried regrouping the parens to make it clearer anyway, thanks

Reply:
was this not viable?

Reply:
as in,
```
if (self.tz is None) or (other.tz is None):
    raise TypeError("Cannot join tz-naive with tz-aware DatetimeIndex")
```
?

I don't think that would work because we don't want to raise if _both_ `self.tz` and `other.tz` are `None`, just if one is but the other isn't

Reply:
Not `(self.tz is None) or (other.tz is None)`, but `(self.tz is None) ^ (other.tz is None)`

Reply:
ah sorry, I didn't actually know that was a Python command - that should work then, thanks so much!